website/project/model.py

```diff
@@ -2609,3 +2738,137 @@ def to_json(self):
+                self.state = 'retracted'
+                # Remove any embargoes associated with the registration
+                parent_registration = Node.find_one(Q('retraction', 'eq', self))
+                if parent_registration.is_embargoed or parent_registration.pending_embargo:
+                    parent_registration.embargo.state = 'cancelled'
+                    parent_registration.embargo.save()
+        except KeyError:
+            raise PermissionsError('User must be an admin to disapprove retraction of a registration.')
+
+
+def validate_embargo_state(value):
+    acceptable_states = ['unapproved', 'active', 'cancelled', 'completed']
+    if value not in acceptable_states:
+        raise ValidationValueError
+    return True
+
+
+class Embargo(StoredObject):
+    """Embargo object for registrations waiting to go public."""
+
+    _id = fields.StringField(primary=True, default=lambda: str(ObjectId()))
+    initiation_date = fields.DateTimeField(auto_now_add=datetime.datetime.utcnow)
+    initiated_by = fields.ForeignField('user', backref='embargoed_by')
+    end_date = fields.DateTimeField()
+    # Expanded: Dictionary field mapping admin IDs their approval status and relevant tokens:
+    # {
+    #   'b3k97': {
+    #     'has_approved': False,
+    #     'approval_token': 'Pew7wj1Puf7DENUPFPnXSwa1rf3xPN',
+    #     'disapproval_token': 'TwozClTFOic2PYxHDStby94bCQMwJy'}
+    # }
+    approval_state = fields.DictionaryField()
+    # One of 'unapproved', 'active', 'cancelled', or 'completed
+    state = fields.StringField(default='unapproved', validate=validate_embargo_state)
+    for_existing_registration = fields.BooleanField(default=False)
+
+    @property
+    def is_embargoed(self):
+        if self.state == 'active':
+            return self.end_date.strftime("%A, %b. %d, %Y")  # e.g. 'Friday, Jan. 01, 2016'
+        return False
+
+    @property
+    def pending_embargo(self):
+        return self.state == 'unapproved'
+
+    # NOTE(hrybacki): Old, private registrations are grandfathered and do not
+    # require to be made public or embargoed. This field differentiates them
+    # from new registrations entering into an embargo field which should not
+    # show up in any search related fields.
+    @property
+    def pending_registration(self):
+        return not self.for_existing_registration and self.pending_embargo
+
+    def disapprove_embargo(self, user, token):
+        """Cancels retraction if user is admin and token verifies."""
+        try:
+            if self.approval_state[user._id]['disapproval_token'] != token:
+                raise InvalidEmbargoDisapprovalToken
+            self.state = 'cancelled'
+            # Delete parent registration if it was created at the time the embargo was initiated
+            if not self.for_existing_registration:
+                parent_registration = Node.find_one(Q('embargo', 'eq', self))
+                parent_registration.is_deleted = True
```

Comment:
That shouldn't be possible as all embargoes are tied to a Node when they are created unless someone is creating one on their own somewhere else.